### ISSUE TITLE
Add partial support for EF migrations on InMemory Provider to enable data migrations and seeding

### ISSUE THREAD

**Issue by Danielku15** [dotnet/efcore#17100]
With this "Feature Request" I'm asking if it could be possible to extend the InMemoryProvider by the necessary functionality to support "data migrations". 

As of today the InMemory provider does not provide EF migrations. This is not a big deal, as long as you only consider schema migrations which are not needed for the InMemoryProvider. But there are also "data migrations" represented by the methods like `InsertData` and `UpdateData` which have a relevance for InMemory databases too. They are currently quite designed for relational databases with their "table" and "column" approach and might need refactoring to be database provider agnostic. 

I have the feeling that none of the data seeding/migration systems (beside a custom built one) works properly with real databases and InMemory databases in parallel. 

We are using data migrations to seed our database and make potentially modifications to the existing data if needed. [The documentation](https://docs.microsoft.com/en-us/ef/core/modeling/data-seeding#model-seed-data) explains a little bit about data seeding but this systems somehow are not working for us. Let me elaborate some scenarios I tried to implement and how I failed: 

### Steps to reproduce

1. Run migrations with data migrations 
We have a migration step where we call `migrationBuilder.InsertData()` to seed some data into the freshly created table right above the `InsertData`. On startup of our (test-)service we call `context.Database.MigrateAsync()` to run the migration chain and get everything prepared. This fails by today with the following exception as the InMemory provider does not support relational operations:

```
System.InvalidOperationException: Relational-specific methods can only be used when the context is using a relational database provider.
    at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.GetRelationalService[TService](IInfrastructure`1 databaseFacade)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.MigrateAsync(DatabaseFacade databaseFacade, CancellationToken cancellationToken)
```

**Expected Behavior:** It would be great if the schema related operations of the migration are simply skipped as they are "not needed" (?) while data migrations should be executed correctly. By implementing the feature request of this issue, we would be able to "migrate" the database. 

**Technical Challenge:** I guess the InMemory provider can only consider the latest snapshot of the model for operations. A problem scenario here is: If `Migration1` inserts some data to `MyEntity` and `Migration2` adds a required property `MyProp` to `MyEntity`. I assume that with the todays implementation, the InMemory provider will fail on the data insert of `Migration1` because the required field information is not filled. I'm not sure how to solve this best, maybe the schema migrations also need to be interpreted in InMemory provider and validations must consider the current state at the time of migration. Maybe validations could be ignored for the sake of in-memory testing. 

2. Use `EnsureCreatedAsync` instead of `MigrateAsync`  as described by the docs. 
The next thing I tried, is to use `EnsureCreatedAsync` as the docs read: 

> Alternatively, you can use context.Database.EnsureCreated() to create a new database containing the seed data, for example for a test database or when using the in-memory provider or any non-relation database.

But this will not work. The `EnsureCreatedAsync` will only try to seed data that was specified via `HasData` via `OnModelCreating` but not `InsertData` statements of any migration. Hence I was left with an empty InMemory database by using `EnsureCreatedAsync` and my tests failed as required entities were not there. 

I guess this is by-design considering the current behavior. Migrations are for relational database, other providers only consider the current model snapshot as per configuration. 

3. Use `HasData` in the `OnModelCreating`
The docs and discussions in the internet (almost) convinced me to use `HasData` for defining the required application entities and ensuring they are inserted. I hope that with defining `HasData` results in EF migrations with data migrations being generated. But I did not came that far to check how it works.

I added an initial data row to one of my entities and instantly got some errors because we are forced to define `IDENTITY` primary key columns while I hoped that they are auto-generated as usual. A bit or [reading](#11776) later, I let myself convince to add a `Id = 1` to my initial data row. Now it seems to work...unless you try to insert a second record for the same entity. In my test I create a second entity, beside the seeded one. And there I get an exception because the InMemory provider does not pick up the "used primary keys" from seeding and tries to give the first added entity also the id 1 : 

```
System.ArgumentException: An item with the same key has already been added.
    at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.EntityFrameworkCore.InMemory.Storage.Internal.InMemoryTable`1.Create(IUpdateEntry entry)
   at Microsoft.EntityFrameworkCore.InMemory.Storage.Internal.InMemoryStore.ExecuteTransaction(IReadOnlyList`1 entries, IDiagnosticsLogger`1 updateLogger)
   at Microsoft.EntityFrameworkCore.InMemory.Storage.Internal.InMemoryDatabase.SaveChangesAsync(IReadOnlyList`1 entries, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.<SaveChangesAsync>d__81.MoveNext()
```

The InMemory provider behaves like a IDENTITY_INSERT in SQL server where a CHECKIDENT is missing.

**Expected Behavior:** If I seed some initial records with `HasData` I either should not be forced to add the primary key column and let the identity generation fill it (like the `InsertData` also does not require it). Or the InMemory provider must detect that I used some IDs and update the internal identity counter to the next available value. Anything else makes this whole feature useless. It is understandable from #11776 that for migrating seeded data correctly the key must be known, otherwise never a `UpdateData` can be scaffolded. But having no mechanism in place to set the identity start values accordingly sounds like a bug. Either this should be auto-detected or at least a configuration value. 

So here I am, trying to get one of the most simple tasks done: inserting some seed data for my application regardless of the database provider user, but I am failing. Either I missed something very obvious or seeding is simply a task not yet supported well by efcore. 

### Further technical details
EF Core version: 2.2.3
Database Provider: Microsoft.EntityFrameworkCore.InMemory and Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Pro 1809 x64
IDE: Rider 2019.2

**Comment by ajcvickers** [comment]
@Danielku15 Thanks for the suggestion. We discussed this, but given that the in-memory provider is inherently non-relational, and that the migrations feature (especially the actual migrations themselves) are inherently relational, attempting to combine the two in this way would be both architecturally and pragmatically very difficult. Even with this, it's not clear that this would be the best way to manage the data--it seems like it would be very easy for the database to get out-of-sync with the model/migrations.
